### ISSUE TITLE
slow creation of dm-wriboost devices

### ISSUE THREAD

**Issue by onlyjob** [akiradeveloper/dm-writeboost#59]
I've noticed that it takes around 30+ seconds to create dm-writeboost device.
Most of this time dm-writeboost actively reads from caching device.
Is this normal/expected?

Not a big problem but when I boot system with 5 dm-writeboost devices it spends over three minutes preparing them... 

Please consider it as feature request for lazy initialisation. Thanks.

**Comment by akiradeveloper** [comment]
> Most of this time dm-writeboost actively reads from caching device.

For the first time, reformatting is done which causes writes to headers over the cache device.
After that, it reads all headers to reconstruct the in-memory structure.

> Is this normal/expected?

Expected. one of the shortcomings of the log-structured method.

> Not a big problem but when I boot system with 5 dm-writeboost devices it spends almost two minutes preparing them...

I think that's not a big problem too. dm-writeboost is often applied to a server system which is really unlikely to reboot frequently. So the start-up time is not a problem.

> Please consider it as feature request for lazy initialisation. Thanks.

Lazy initialization is probably impossible technically but I have some idea for fast initialization. Because the state of the headers in reformatted cache device is given, we can skip replaying the logs. Another optimization is reformatting can be done outside the kernel.

FYI, fast initialization exhibits the problem especially when we need to swap two cache devices for longevity reason. I recommend users to use `pvmove` given that the two cache device is of the same size. But if he/she needs to upgrade cache device to different size is bit difficult. Here we need to initialize cache device immediately to lower downtime.

**Comment by disaster123** [comment]
Does update_sb_record_interval  help in this case?

**Comment by onlyjob** [comment]
No, I think it does not help. I'm not resuming devices but initialise 'em from scratch with zeroed first block on caching device. Besides "create-n-messasge" style used by writeboosttab do not allow me to pass `update_sb_record_interval` on `dmsetup create`...

**Comment by akiradeveloper** [comment]
If you set update_sb_record_interval to some value other than zero, (at least theoretically) you can get faster resuming.

But the effect isn't noticeable. The reason is the implementation is sub-optimal for faster resuming.

I don't recommend to use update_sb_record_interval because

* It's not tested well.  
* It has side-effect that the resumed cache blocks will be fewer because the algorithm can shorten the resume time by omitting reading cache blocks that has been already written back. As the result, read hit ratio will be down after reboot.

update_sb_record_interval can bring you two effects. One is reducing the resume time and the another is omitting writeback twice after reboot. The implementation for the latter is optimal.

For production use, I recommend to clear cache state on every reboot as Dmitry does. Faster resuming is quite low priority for me. The nob is almost deprecated.

Rebooting storage server is quite infrequent so please take the air while the server is rebooting.

**Comment by disaster123** [comment]
Isn't it possible to do something like:
1.) drop / clear cache on shutdown like Dmitry does
2.) mark cache device as clean
3.) "dmsetup create" succeeds immediatly if cache device is marked as clean

**Comment by akiradeveloper** [comment]
No. It's not possible to immediately return.

Initialization needs two steps (please read `resume_cache`)

1) formatting the cache device (actually, zeroing out all `segment_header` part)
2) reading all the segments

We can know cache device is clean by seeing the first sector of the cache device (that's we zeroed out on shutdown). This is quite fast but we can't return before finishing 1). Without 1), because old segment header remains, we can't survive sudden power failure.

After 1), now that we know the cache device was clean, the state (or dump) of the cache device is fixed, we can run into special path that omits 2).

----

Do you have strong request for this enhancement?

**Comment by akiradeveloper** [comment]
Hi Stefan,

I've just experimentally implemented fast-path for first (or clean) creation.

It's published as fast-replay branch. So could you please try it?

It's very helpful if you measure the creation time between v2.0.3 and fast-replay in case of clean creation. I believe it's quite improved.

Fortunately, I've got no sign of failure in my regression tests. So, if you like the improvement I will immediately merge it into master and ship as v2.0.4.

**Comment by disaster123** [comment]
Works absolutely GREAT! It now maps in 10s instead of 7minutes for my 90GB SSD.

Thanks a lot!

**Comment by akiradeveloper** [comment]
Oh... Great

So I will fold this enhancement into v2.0.4

**Comment by onlyjob** [comment]
Awesome. 2.0.4 initialises so much faster. Thank you. :)

**Comment by akiradeveloper** [comment]
I am glad to hear that. I think this enhancement makes your script more valuable.

**Comment by onlyjob** [comment]
I think this enhancement makes _dm-writeboost_ more usable. ;)

**Comment by disaster123** [comment]
Yes.
@onlyjob
i've modified your writeboost script to map and unmap in parallel? Would you like to have the patch?
and i've modified the init script to run very early at boot to be able to mount those devices through fstab and unmap very lately so that the devices are already unmapped.

**Comment by onlyjob** [comment]
Of course I want patch. :) We want to make this software as usable and as useful as we can, aren't we? ;)
@disaster123, when you map in parallel how do you make sure all mappings are completed? I reckon I just need to have a look at your patch. Anyway with 2.0.4 mapping of new devices (i.e. no resuming) take only few seconds.
I'm curious about your modifications to init script. Please feel free to open a ticket or merge request on GitLab: https://gitlab.com/onlyjob/writeboost
Thanks.
